File: AllSorts/allSorts.js
function quickSort(array, lo, hi) {
	if(lo < hi) {
		var pivot = partition(array, lo, hi);
		quickSort(array, lo, pivot-1);
		quickSort(array, pivot+1, hi);
	}
	return array;
}

function partition(array, lo, hi) {
	var pivot = array[hi];
	i = lo;
	for(var j = lo; j < hi; j++) {
		if(array[j] <= pivot) {
			swap(array,i,j);
			i++; // move the wall up 1 place
		}
	}
	swap(array,i,hi);
	return i;
}

function bubbleSort(array) {
	var swapped = true;
	while (swapped == true) {
		swapped = false;
		for(var i = 1; i < array.length; i++) {
			if (array[i-1] > array[i]) {
				swap(array, i-1, i);
				swapped = true;
			}
		}
	}
	console.log("Bubble sorted:", array);
	return array;
}

function selectionSort(array) {
	for(var i = 0; i < array.length; i++) {
		var minimum = i;
		for(var j = i+1; j < array.length; j++) {
			if(array[j] < array[minimum]) {
				minimum = j;
			}
		}
		swap(array,i,minimum);
	}
	console.log("Selection sorted:", array);
	return array;
}

function insertionSort(array) {
	for(var i = 1; i < array.length; i++) {
		var j = i;
		while(array[j-1] > array[j] && j > 0) {
			swap(array,j-1,j);
			j--;
		}
	}
	console.log("Insertion sorted:", array);
	return array;
}

function mergeSort(array, lo, hi, workArray) {
	if(array.length < 2) {
		return array;
	} else {
		var mid = Math.floor((hi+lo) / 2);
		mergeSort(array, lo, mid, workArray);
		mergeSort(array, mid+1, hi, workArray);
		merge(array, lo, mid, hi, workArray);
		CopyArray(B, lo, hi, A);
	}

}

function merge(A, lo, mid, hi, B) {
	var i = lo, j = mid;
	for(var k = lo; k < hi; k++) {
		if(i < mid && (j >= hi || array[i] <= array[j])) {
			B[k] = A[i];
			i++;
		} else {
			B[k] = A[j];
			j++;
		}
	}
}

function CopyArray(workArray, lo, hi, array) {
	for(var k = lo; k < hi; k++) {
		array[k] = workArray[k];
	}
}

function swap(array, i, j) {
	var temp = array[i];
	array[i] = array[j];
	array[j] = temp;
}

var unsorted = [4,3,2,1,6];
document.getElementById("sel").innerHTML = selectionSort(unsorted);

File: AllSorts/allsorts.py
'''
Created on 23 Sep 2015

@author: tobydobbs
'''

import math

def swap(array, i, j):
    temp = array[i]
    array[i] = array[j]
    array[j] = temp
    
def insertionSort(array):
    for i in range(len(array)):
        j = i
        while j > 0 and array[j-1] > array[j]:
            swap(array, j, j-1)
            j = j - 1

def bubbleSort(array):
    swapped = True
    while(swapped == True):
        swapped = False
        for i in range(len(array)-1):
            if array[i] > array[i+1]:
                swap(array, i, i+1)
                swapped = True
    return array

def selectionSort(array):
    for i in range(len(array)):
        minimum = i
        for j in range(i+1, len(array)):
            if array[j] < array[minimum]:
                minimum = j
        swap(array, i, min)
        
def quickSort(array, lo, hi):
    if lo < hi:
        p = partition(array, lo, hi)
        quickSort(array, lo, p - 1)
        quickSort(array, p + 1, hi)

def partition(array, lo, hi):
    wall = lo
    pivot = hi
    for i in range(lo, hi):
        if array[i] < array[pivot]:
            swap(array, wall, i)
            wall += 1 # move the wall up one space
    swap(array, wall, hi)
    return wall

def mergeSort(array):
    print "array: " + str(array)
    if len(array) == 1:
        return array
    left, right, middle = [], [], int(math.floor(len(array)/2))
    for i in range(middle):
        left.append(array[i])
    for i in range(middle, len(array)):
        right.append(array[i])
        
    left = mergeSort(left)
    right = mergeSort(right)
    
    return merge(left, right)

def merge(left, right):
    print "left: " + str(left) + ", right: " + str(right)
    result = []
    while len(left) and len(right):
        if left[0] <= right[0]:
            result.append(left[0])
            del left[0]
        else:
            result.append(right[0])
            del right[0]
    while len(left):
        result.append(left[0])
        del left[0]
    while len(right):
        result.append(right[0])
        del right[0]
    return result

def main():
    array = [1,5,6,3,9,4,6,2,4,5,3,8]
    print array
    quickSort(array, 0, len(array) - 1)
    print array
    
if __name__ == "__main__":
    main()      
